### PR TITLE
Support reading configuration from etcd

### PR DESCRIPTION
**Pull request by mbarnes**
See https://github.com/projectatomic/commissaire/issues/135 for rationale.

`read_config_file()` will now first try to read configuration from `etcd`, if an environment variable named `"ETCD_MACHINES"` is defined (a comma-delimited list of URLs).

### CODE REVIEW

**Comment by mbarnes** [comment]
@ashcrow Code's done if you want to review it.

Keeping this **WIP** a little longer while I update documentation.

**Comment by mbarnes** [comment]
:arrow_up:

**Comment by mbarnes** [comment]
[(5) [multi-host-mgr-2] Containerize individual Commissaire services](https://trello.com/c/ZsuDHbez/738-5-multi-host-mgr-2-containerize-individual-commissaire-services)

File: setup.cfg [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [flake8]
 ignore = E402
 exclude = test/*
-max-complexity = 15
+max-complexity = 17
```
Comment (ashcrow) [inline setup.cfg:4]: Do you know what area caused the bump up?

File: src/commissaire/util/config.py [hunk 2]
```diff
@@ -51,9 +57,92 @@ def _normalize_member_names(json_object):
...
+    """
+    args = {
+        'cert': None,  # see below
+        'ca_cert': os.environ.get('ETCD_CACERT'),
+        'username': os.environ.get('ETCD_USERNAME'),
+        'password': os.environ.get('ETCD_PASSWORD')
+    }
+
+    try:
+        args['cert'] = (
```
Comment (ashcrow) [inline src/commissaire/util/config.py:82]: nit: May want to throw some ``logging.debug`` statements in. I can forsee someone starting a server up and things not working and having those statements in silent ``pass``es may be helpful.

File: src/commissaire/util/config.py [hunk 3]
```diff
@@ -71,27 +160,34 @@ def read_config_file(path=None, default=C.DEFAULT_CONFIGURATION_FILE):
...
-            '{}: File content must be a JSON object'.format(path))
+    if 'ETCD_MACHINES' in os.environ:
+        # Derive etcd key from the default file path.
+        basename = os.path.basename(default).split('.', 1)[0]
+        key = '/commissaire/config/' + basename
+        _read_etcd_config_key(key, json_object)
+
+    if not json_object:
+        # As with the fileinput module, replace '-' with sys.stdin.
+        if path == '-':
```
Comment (ashcrow) [inline src/commissaire/util/config.py:188]: 👍 good idea

File: test/test_util_config.py [hunk 4]
```diff
@@ -119,6 +121,62 @@ def test_read_config_file_with_invalid_authentication_plugin(self):
...
+        """
+        Verify _read_etcd_config_key handles ETCD environment variables.
+        """
+        environ = {
+            'ETCD_MACHINES': 'http://localhost:4001,'
+                             'http://etcd.example.com:4001',
+            'ETCD_TLSKEY': '/path/to/cert.key',
+            'ETCD_TLSPEM': '/path/to/cert.pem',
+            'ETCD_CACERT': '/path/to/cacert.pem',
+            'ETCD_USERNAME': 'neo',
```
Comment (ashcrow) [inline test/test_util_config.py:134]: neo

**Review by ashcrow — APPROVED** [review]
👍 A few minor suggestions. At a later date we should try to improve coverage on ``storage/etcd.py``

```
commissaire/storage/etcd.py                          79     61    23%
```

File: setup.cfg (hunk 1 above)
Comment (mbarnes) [reply setup.cfg:4]: Yeah, `read_config_file()` is pushing the complexity limit.

I split off the helper function `_read_etcd_config_key()` to try and keep the complexity under control, but even then the small
```
if 'ETCD_MACHINES' in os.environ:
    ...
```
block I added bumped the complexity a couple points.

Was gonna circle back later and try to break it up further in a separate cleanup PR.

File: src/commissaire/util/config.py (hunk 2 above)
Comment (mbarnes) [reply src/commissaire/util/config.py:82]: Good idea.  Will do.

**Comment by mbarnes** [comment]
> At a later date we should try to improve coverage on storage/etcd.py

Don't disagree, but in general I think it's gonna be hard to unit test backends for external services without mocking it out to such an extent that there's no real logic left to test.  I suggest relying more on the E2E tests (through vagrant, if necessary) for that.  We already have good coverage of `etcd.py` there.

**Comment by mbarnes** [comment]
:arrow_up:
Note, logging isn't set up yet (since it's part of the JSON configuration) so I'm just printing variables to `stdout` if the environment defines at least `ETCD_MACHINES`.

**Comment by ashcrow** [comment]
👍

**Comment by ashcrow** [comment]
Manually merged.

**Pull request closed without merging**